tools/deploy/helper.go

```diff
@@ -0,0 +1,434 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package deploy
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/services/placement"
+	"github.com/m3db/m3x/errors"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+	"github.com/m3db/m3x/sync"
+)
+
+var (
+	errNoDeploymentProgress = errors.New("no deployment progress")
+	errInvalidRevision      = errors.New("invalid revision")
+)
+
+// Mode is the deployment mode.
+type Mode int
+
+// A list of supported deployment modes.
+const (
+	DryRunMode Mode = iota
+	ForceMode
+)
+
+// Helper is a helper class handling deployments.
+type Helper interface {
+	// Deploy deploys a target revision to the instances in the placement.
+	Deploy(revision string, mode Mode) error
+}
+
+// TODO(xichen): disable deployment while another is ongoing.
```

Comment:
Not fully, if you run the tool while another one is ongoing and is in between steps, you might still be able to proceed.